addOns/graphql/src/main/java/org/zaproxy/addon/graphql/GraphQlSpider.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2020 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.addon.graphql;
+
+import java.util.Locale;
+import net.htmlparser.jericho.Source;
+import org.apache.commons.lang.StringUtils;
+import org.apache.log4j.Logger;
+import org.parosproxy.paros.network.HttpHeader;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpSender;
+import org.zaproxy.zap.spider.parser.SpiderParser;
+
+public class GraphQlSpider extends SpiderParser {
+
+    private static final Logger LOG = Logger.getLogger(GraphQlSpider.class);
+    private String schema;
+
+    @Override
+    public boolean parseResource(HttpMessage message, Source source, int depth) {
+        try {
+            GraphQlParser parser =
+                    new GraphQlParser(
+                            message.getRequestHeader().getURI(), HttpSender.SPIDER_INITIATOR);
+            if (schema != null && !schema.isEmpty()) {
+                parser.parse(schema);
+            } else {
+                parser.introspect();
+                parser.persistImportMessage();
+            }
+        } catch (Exception e) {
+            LOG.debug(e.getMessage(), e);
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean canParseResource(HttpMessage message, String path, boolean wasAlreadyConsumed) {
+        try {
+            String uri = message.getRequestHeader().getURI().toString();
+            String contentType =
+                    message.getResponseHeader()
+                            .getHeader(HttpHeader.CONTENT_TYPE)
+                            .toLowerCase(Locale.ROOT);
+            String responseBodyStart =
+                    StringUtils.left(message.getResponseBody().toString(), 250)
+                            .toLowerCase(Locale.ROOT);
+            if (uri.endsWith(".graphql")
+                    || uri.endsWith(".graphqls")
+                    || contentType.startsWith("application/graphql")) {
+                schema = message.getResponseBody().toString();
+                LOG.debug("Found GraphQL schema at: " + uri);
```

Comment:
Should return true? Or the following check is done to ensure?

Reply:
If we only have the schema, that is of no use to us without an endpoint. So the spider returns true only if we have an endpoint.

Reply:
I see, thanks for the clarification.

Reply:
np